infer/provider.go

```diff
@@ -62,16 +63,28 @@ func NewProvider() *Provider {
 
 // Retrieve the configuration of this provider.
 //
-// Note: Config will panic if the type of T does not match the type of the config or if
+// Note: GetConfig will panic if the type of T does not match the type of the config or if
 // the provider has not supplied a config.
 func GetConfig[T any](ctx p.Context) T {
-	cv := ctx.Value(configKey)
+	v := ctx.Value(configKey)
+	return getConfig[T](v, v != nil)
+}
+
+// Retrieve the configuration of this provider.
+//
+// Note: GetComponentConfig will panic if the type of T does not match the type of the config or if
+// the provider has not supplied a config.
+func GetComponentConfig[T any](ctx *pulumi.Context) T {
```

Comment:
In addition to being able to access the provider config from component factories, I could imagine component authors wanting to access other provider-related information from the `p.Context` (e.g. being able to call the `Log` methods, etc.).

Should we provide a way to convert a `*pulumi.Context` from `Construct` into a `p.Context`? Effectively a function like `p.CtxFromPulumiContext(ctx *pulumi.Context) p.Context` that would live next to the other `Ctx` functions?

Then we could choose to not expose `GetComponentConfig` if we wanted. We could just have `GetConfig`, and component authors would have to do something like:

```go
// Get provider config.
pctx := p.CtxFromPulumiContext(ctx)
config := infer.GetConfig[Config](pctx)

// And do other things with the provider context.
pctx.Log(...)
```

If we did still want a helper like `GetComponentConfig` we could include it and it could be implemented like:

```go
func GetComponentConfig[T any](ctx *pulumi.Context) T {
	pctx := p.CtxFromPulumiContext(ctx)
	return GetConfig[T](pctx)
}
```

Reply:
I think this is good future-proofing. 

You have access to `pulumi.Context.Log` already, so I don't think this is necessary for logging.